src/gpuarray_collectives_cuda_nccl.c

```diff
@@ -253,7 +253,7 @@ static int reduce(gpudata* src, size_t offsrc, gpudata* dest, size_t offdest,
   GA_EXIT_ON_ERROR(ctx, cuda_wait(src, CUDA_WAIT_READ));
   // sync: wait till a read/write has finished (out of concurrent kernels)
   if (rank == root)
-    GA_EXIT_ON_ERROR(ctx, cuda_wait(dest, CUDA_WAIT_WRITE));
+    GA_EXIT_ON_ERROR(ctx, cuda_wait(dest, CUDA_WAIT_ALL));
```

Comment:
CUDA_WAIT_WRITE does record to `wev` already.

However, we need to wait for previous writes as you say.

If you make this change, please do it in a separate PR, this one is big enough :)